### PR TITLE
Introduce ConstSymbolData for const references to Symbols.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Introduce ConstSymbolData for const references to Symbols.

`const SymbolData` is trivially convertable to a `SymbolData`, which permits mutations to `Symbol`s even with a `const GlobalState&`!

This change locks that down so that this escape hatch is not viable. Instead of relying on `const SymbolData` to protect the symbol against mutation, introduce a new class called `ConstSymbolData`.

It includes an explicit `const_cast` to permit code that relied on this behavior until it can be fully fixed.

### Motivation

Prevent code from mutating symbol table with a `const GlobalState`.

### Test plan

Compiles and tests pass.

### CODE REVIEW

File: core/Symbols.cc [hunk 1]
```diff
@@ -1400,7 +1400,9 @@ vector<std::pair<NameRef, SymbolRef>> Symbol::membersStableOrderSlow(const Globa
     return result;
 }
 
-SymbolData::SymbolData(Symbol &ref, const GlobalState &gs) : DebugOnlyCheck(gs), symbol(ref) {}
+SymbolData::SymbolData(Symbol &ref, GlobalState &gs) : DebugOnlyCheck(gs), symbol(ref) {}
+
+ConstSymbolData::ConstSymbolData(const Symbol &ref, const GlobalState &gs) : DebugOnlyCheck(gs), symbol(ref) {}
```
Comment (jez) [inline core/Symbols.cc:1405]: So this is the magic, right? It's a type error to construct a `SymbolData` if you don't have a `const GlobalState&`?

File: resolver/resolver.cc [hunk 2]
```diff
@@ -1418,7 +1419,8 @@ class ResolveTypeMembersWalk {
             // >   Boolean = T.let(nil, T.untyped)
             // > end
             if (data->isTypeAlias() && send->fun == core::Names::let()) {
-                data->resultType = core::Types::untypedUntracked();
+                // TODO(jvilk): Fix this.
+                const_cast<core::TypePtr &>(data->resultType) = core::Types::untypedUntracked();
```
Comment (jez) [inline resolver/resolver.cc:1423]: 😱

File: resolver/resolver.cc [hunk 3]
```diff
@@ -1418,7 +1419,8 @@ class ResolveTypeMembersWalk {
             // >   Boolean = T.let(nil, T.untyped)
             // > end
             if (data->isTypeAlias() && send->fun == core::Names::let()) {
-                data->resultType = core::Types::untypedUntracked();
+                // TODO(jvilk): Fix this.
```
Comment (jez) [inline resolver/resolver.cc:1422]:
```suggestion
                // TODO(jvilk): This mutates the symbol table from an immutable context due to a const bug in Symbols.cc. DO NOT COPY THIS
```

**Review by jez — APPROVED** [review]

File: core/Symbols.cc (hunk 1 above)
Comment (jvilk-stripe) [reply core/Symbols.cc:1405]: The magic is that you must now have a non-const `GlobalState` to construct a `SymbolData`.

Previously, you could construct a `const SymbolData` with a `const GlobalState`. However, you can trivially copy a `const SymbolData` into a `SymbolData` and have mutable access to the symbol. It turns out that the compiler would do this transformation for you with the `auto` keyword, which is some bad stuff.

Comment (jez) [reply core/Symbols.cc:1405]: So if we did `const auto symData` everywhere it would work?

Comment (jvilk-stripe) [reply core/Symbols.cc:1405]: More-or-less. You could still accidentally copy it into a non-const `SymbolData` but we don't typically use `SymbolData` objects like that.

**Pull request merged**